File: 826.most-profit-assigning-work.py
#
# @lc app=leetcode id=826 lang=python3
#
# [826] Most Profit Assigning Work
#

# @lc code=start
class Solution:
    def maxProfitAssignment(self, difficulty: List[int], profit: List[int], worker: List[int]) -> int:
        tasks = list(zip(difficulty, profit))
        tasks.sort()
        res = 0
        ind = 0
        max_profit = 0
        for w in sorted(worker):
            while ind < len(tasks) and w >= tasks[ind][0]:
                max_profit = max(max_profit, tasks[ind][1])
                ind += 1
            res += max_profit
        return res
# @lc code=end

File: 826.most-profit-assigning-work.cpp
/*
 * @lc app=leetcode id=826 lang=cpp
 *
 * [826] Most Profit Assigning Work
 */
#include <vector>
#include <queue>
#include <algorithm>
using namespace std;
// @lc code=start
class Solution
{
public:
    int maxProfitAssignment(vector<int> &difficulty, vector<int> &profit, vector<int> &worker)
    {
        auto comp = [](pair<int, int> &l, pair<int, int> &r)
        {
            if (l.first < r.first)
            {
                return true;
            }
            else if (l.first == r.first)
            {
                return l.second > r.second;
            }
            else
            {
                return false;
            }
        };
        priority_queue<pair<int, int>, vector<pair<int, int>>, decltype(comp)> q(comp);
        int width = difficulty.size();
        for (int i = 0; i != width; ++i)
        {
            q.push({profit[i], difficulty[i]});
        }
        sort(worker.begin(), worker.end(), greater<int>());
        int res = 0;
        int ind = 0;
        int worker_num = worker.size();
        while (ind != worker_num && !q.empty())
        {
            if (worker[ind] >= q.top().second)
            {
                res += q.top().first;
                ++ind;
            }
            else
            {
                q.pop();
            }
        }
        return res;
    }
};
// @lc code=end
